gh-46771: docs for asyncio.timeout (#94764)
python/cpython: Doc/library/asyncio-task.rst

      Added the *context* parameter.

Task Cancellation
=================

Tasks can easily and safely be cancelled.
When a task is cancelled, asyncio will raise a :exc:`asyncio.CancelledError`

> into the task at the next opportunity.

Reviewer: I may just be a n00b, but what exactly does it mean to raise an error _into_ a task? What determines when there will be an "opportunity"?

> Coroutines are recommended to use ``try/finally`` blocks to robustly

Reviewer 2: The grammar here gets my hackles up a bit but I'm at a loss for how to put it better. (@CAM-Gerlach ?)

Reviewer: Yeah, not only that but it is also potentially ambiguous (at least to a relative asyncio n00b like myself), as "coroutines" can be parsed either as the object of the sentence (i.e., that they are the target of the recommendation, which I'm guessing is the intended meaning), or that they are recommended _in order to use_ ``try/finally`` blocks.

Assuming the former, what about just

```suggestion
It is recommended that coroutines use ``try/finally`` blocks to robustly
```

Sure, it uses passive voice, but IMO its a good deal clearer

perform clean-up logic. In the case the :exc:`asyncio.CancelledError`
is explicitly caught, it should generally be propagated when
clean-up is complete. Most code can safely ignore :exc:`asyncio.CancelledError`.

Important asyncio components, like :class:`asyncio.TaskGroup` and the
:func:`asyncio.timeout` context manager, are implemented using cancellation

> internally.

Reviewer 2: Maybe add "and might misbehave if a couroutine swallows `CancelledError`"?

Timeouts
========

.. coroutinefunction:: timeout(delay)

    A convenient way to limit the amount of time spent waiting on
    something is to use the :func:`asyncio.timeout`
    :ref:`asynchronous context manager <async-context-managers>`.

>     *delay* can either be ``None`` or a float or int number of

Reviewer: ```suggestion
    *delay* can either be ``None``, or a float/int number of
```

Avoid awkward and somewhat difficult to parse grammar

    seconds to wait. If *delay* is ``None``, no time limit will

>     be applied. In either case, the context manager can be

Reviewer 2: Maybe explain why supporting `delay=None` is useful.

>     rescheduled after creation.

Reviewer 2: Explain how to reschedule.

    Example::

        async def main():
            async with asyncio.timeout(10):
                await long_running_task()

    If ``long_running_task`` takes more than 10 seconds to complete,
    the context manager will cancel the current task and handle
    the resulting :exc:`asyncio.CancelledError` internally, transforming it

>     into a :exc:`asyncio.TimeoutError` which can be caught and handled.

Reviewer: ```suggestion
    into an :exc:`asyncio.TimeoutError` which can be caught and handled.
```

Typo

    .. note::

      The :func:`asyncio.timeout` context manager is what transforms
      the :exc:`asyncio.CancelledError` into an :exc:`asyncio.TimeoutError`,
      which means the :exc:`asyncio.TimeoutError` can only be caught
      *outside* of the context manager.

    Example of catching :exc:`asyncio.TimeoutError`::

        async def main():
            try:
                async with asyncio.timeout(10):
                    await long_running_task()
            except TimeoutError:
                print("The long operation timed out, but we've handled it.")

            print("This statement will run regardless.")

    The context manager produced by :func:`asyncio.timeout` can be

>     rescheduled to a different deadline and inspected.

Reviewer 2: This should use a new function heading so it's clear that this is about `reschedule` (and that gets added to the index). You may actually have to change the structure of the whole section to admit the existence of the `Timeout` class so you can document its methods (aren't there a few more worth documenting?)
